hkmc2/shared/src/test/mlscript/nofib/mate.mls

```diff
@@ -0,0 +1,652 @@
+    ms,
+    rawmoves(c, ksq, bd)
+  ),
+  Nil,
+  forcesColoured(c, bd)
+)
+
+fun comment(s) = (s is Nil) || listEq(take(2, s), nofibStringToList("--"))
+
+fun last(ls) = if ls is
+  x :: Nil then x
+  h :: t then last(t)
+
+fun intOfString(s) = globalThis.parseInt(nofibListToString(s))
+
+fun parseGoal(ls) = if ls is
+  gltxt :: Nil then
+    let ws = words(gltxt)
+    let c = if listEq(head(ws), nofibStringToList("Black")) then Black else White
+    let n = intOfString(last(ws))
+    [c, n]
+  else throw Error("parseGoal")
+
+fun parseSquare(r, f, c) = if c === "-" then Nil else
+  let clr = if isUpper(c) then Black else White
+  let kin = if toLower(c) ===
+    "k" then King
+    "q" then Queen
+    "r" then Rook
+    "b" then Bishop
+    "n" then Knight
+    "p" then Pawn
+  [[clr, kin], [f, r]] :: Nil
+
+fun parseRank(r, x) = concat(zipWith(
+  (a, b) => parseSquare(r, a, b),
+  enumFromTo(1, 8),
+  filter(pp => not(pp === " "), x)
+))
+
+
+fun parseBoard(ls) =
+  fun addPiece(p_sq, x) = if p_sq is [p, sq] then putPieceAt(sq, p, x)
+  foldr(
+    addPiece,
+    emptyBoard,
+    concat(zipWith(parseRank, reverse(enumFromTo(1, 8)), ls))
+  )
+
+fun parseProblem(s) =
+  let bdtxt_gltxt = splitAt(8, filter(x => not(comment(x)), s))
+  if bdtxt_gltxt is [bdtxt, gltxt] then
+    let bd = parseBoard(bdtxt)
+    let gl = parseGoal(gltxt)
+    [bd, gl]
+
+fun readProblem(s) = parseProblem(lines(s))
+
+:...
+//│ ————————————————————————————————————————————————————————————————————————————————
+fun replies(bd, c, n) =
+  let mds = moveDetailsFor(c, bd)
+  fun solnAnd(mifb, rest) = if mifb is [mif, b] then
+    let sm = solution(b, opponent(c), n - 1)
+    if sm is
+      None then None
+      Some(s) and
+        rest is
+          None then None
+          Some(ms) then Some([mif, s] :: ms)
+  if n
+    === 0 and
+      null_(mds) then Some(Nil)
+      else None
+    > 0 then foldr(solnAnd, Some(Nil), mds)
+    else throw Error("n < 0")
+
+fun solution(bd, c, n) =
+  fun solnOr(mifb, other) = if mifb is
+    [mif, b] then
+      let rsm = replies(b, opponent(c), n - 1)
+      if rsm is
+        None then other
+        Some(Nil) and
+          kingincheck(opponent(c), b) then
+            Some(Solution(mif, Nil))
+          else other
+        Some(rs) then Some(Solution(mif, rs))
+  if n > 0 then
+    let mds = moveDetailsFor(c, bd)
+    foldr(solnOr, None, mds)
+  else throw Error("n <= 0")
+//│ ————————————————————————————————————————————————————————————————————————————————
+
+class Soln(a: MoveInFull, b: List[Tup2[List[MoveInFull],Soln]])
+
+fun polyGt(a, b) =
+  // print(a + " > " + b + " = " + (a > b))
+  a > b
+
+// fun polyLt(a, b) = a < b
+
+fun tab(n) = if n <= 0 then Nil else " " :: tab(n - 1)
+
+:...
+//│ ————————————————————————————————————————————————————————————————————————————————
+fun showReplies(rs, n) = if rs is
+  Nil then Nil
+  [mifs, s] :: rs then
+    append(
```

Comment:
I suggest using `of` again here.

Reply:
And similarly in other places of the file.

Reply:
Thanks for the suggestion! For all the list appension operations in the latest commit I now use an infix operator `+:`, maybe that seems cleaner.

Reply:
Indeed much cleaner!